### ISSUE TITLE
[CV_STN] Spatial Transformer Networks

### ISSUE THREAD

**Issue by jeonggg119** [jeonggg119/DL_paper#3]
Spatial Transformer Networks

## 0. Abstract
- **CNN** : limited by lack of ability to be 'spatially invariant' to input data
- **STN** : Spatially Transform data within Network without extra training supervision
  - can be  inserted into conv architectures
  - invariant to translation, scale, rotation, generic warping, etc
  > spatial invariance : 이미지가 변환되어도 그 이미지로 인식하는 것

## 1. Introduction
- **CNN** Limitation : local max-pooling(2 x 2) help but intermediate feature maps still not invariant to global transformation of input 
  - **Pooling layer** (fixed and local receptive fields) : limited, pre-defined mechanism for dealing with variations
- **Spatial Transformer module** :  dynamic mechanism -> appropriate transformation for each input data on entire feature map (non-locally)
  - select most relevant regions (attention) & transform them to canonical pose
  - can be trained with standard backprop -> end-to-end training
- **STN (CNN + ST module)** 3 Benefits
  - **image classification** : crop, scale-normalization -> simplify subsequent classification task -> great performance
  - **co-localisation** : localize different instances of the same but unknown class
  - **spatial attention** (select most relevent region) : more flexible and trained within backprop without reinforcement learning 

## 2. Related Work (prior work)
- **modeling** transformation with NN 
  - Hinton : 2D affine transformation -> generative model training
  - Tieleman : generative capsule models -> learn discriminative features for classification 
- **transformation-invariant** representation
  - Cohen & Welling : G-CNN
  - Scattering networks, Filter banks
   > Filter Bank : an array of bandpass filters that separates the input signal into multiple components
- attention and detection mechanism for **feature selection**
- STN : invariant representation by manipulating **data** ! (feature extractor X)

## 3. Spatial Transformers
![image](https://user-images.githubusercontent.com/83633885/136578258-086bd4d9-7c92-4571-81df-b969ed86bc0f.png)

- Spatial Transformer = Localisation net + Grid generator + Sampler
- (1) **Input feature map U**가 **Localisation net**에 들어가 transformation parameter θ를 뽑아냄
- (2) θ가 **Grid generator**에 들어가 Sampling point가 지정된 Sampling grid T_θ(G)를 생성함
- (3) **Sampler**에는 Input feature map U와 Sampling grid T_θ(G)가 입력됨
- (4) Sampling grid T_θ(G)에는 Sampling point가 찍혀있기에 이를 U에 적용하면 **Output feature map V**를 얻을 수 있음

### 3.1 Localisation Network
- Regress θ automatically to improve overall accuracy 
- **input** : input feature map U (Width x Height x Channel) 
- **output** : transform parameter θ = f_loc(U)
  - θ : parameter matrix -> shape은 transformation type 따라 달라짐 (ex. affine : 6d)
- **f_loc( )** can take any form (ex. fc net or conv net) BUT should include final Regression layer to produce θ

### 3.2 Parameterized Sampling Grid (Grid generator)
![image](https://user-images.githubusercontent.com/83633885/136598658-3bc1623d-afbf-4903-abee-bef65fd2c6d1.png)

- Generate coordinate grid on input image corresponding to each pixel from output image
- Regular Grid : G = {Gi} of pixels Gi = (xi^t, yi^t) <- output feature map grid
- **Grid generator** : **T_θ( )** can have any differentiable parameterized form 
  - **ex 1)** 2D affine transformation (crop, translation, rotation, scale, skew) matrix -> by 6 params
  - **ex 2)** Attention (crop, translation, isotrophic scaling) : more constrained(=low complexity) -> by 3 params
  - **ex 3)** plane projective (8 params), wise affine, thin plate spline 

- height and width normalized coordinates 
  - (xi^s, yi^s) : source coordinates in the input feature map U 
  - (xi^t, yi^t) : target coordinates in the output feature map V

![image](https://user-images.githubusercontent.com/83633885/136675653-d186be87-ccf8-4767-aa47-9163995fed76.png)

### 3.3 Differentiable Image Sampling (Sampler)
![image](https://user-images.githubusercontent.com/83633885/159680872-877ce55d-17cf-4957-9dc6-2838931d5fa9.png)

![image](https://user-images.githubusercontent.com/83633885/136598703-1e054824-a579-49ae-96a6-e37465ba4beb.png)

- **Input** :  Apply set of sampling points **T_θ(G)** to input image **U** -> Define spatial location in the input 
  - Unm^c : input value at (n,m) in channel c
- **Output** : transformed output feature map **V** (Width x Height x Channel) 
  - Vi^c : output value for pixel i at (xi^t, yi^t) in channel c
- **k( )** can take any sampling kernel as long as (sub-)gradients can be defined (ex. bilinear interpolation) for backprop
  - **(1) Nearest integer** : not-differentiable
    ![image](https://user-images.githubusercontent.com/83633885/159683731-86dee2d8-aab5-44a8-b015-9d357f9717d9.png)
  - **(2) Bilinear interpolation** : sub-differentiable !
    ![image](https://user-images.githubusercontent.com/83633885/159683843-c5b213cc-3216-4148-b061-2fc195fbdf90.png)
    - U와 G에 대해 미분 가능하면 backprop 가능
    - 미분불가능한 구간 있더라도 구간별로 나눠서 backprop 하면 되므로 문제 없음
![image](https://user-images.githubusercontent.com/83633885/159684155-bab27f7b-7a43-4481-9a09-50c349cc7d64.png)

- **Spatial Consistency** : Sampling is done identically for each channel -> every channel is transformed in identical way
  - 같은 input의 다른 channel에 대해선 당연히 같은 sampling

### 3.4 Spatial Transformer Networks
- Spatial Transformer module = Localisation net + Grid generator + Sampler
- **STN** = CNN + ST module (at any point, in any number)
  - Eval 결과, CNN 입력 바로 앞에 ST layer 두는게 일반적으로 가장 효과적
- **Advantages**
  - Fast & little overhead naively & even speedups in attentive models
  - Minimize overall cost during training -> little effect on speed
    - training 과정 중에 모델의 다른 파라미터들과 함께 학습이 되기에 속도에 미치는 영향 거의 X
  - How to transform each sample is compressed in weights of localisation net during training
  - Possible to Downsample or Oversample feature map 
  - Possible to have Multiple spatial transformers in CNN
    - At increasing depths of CNN -> more abstract representations
    - For localisation networks -> more informative representations to base predicted params
    - Parallel -> useful to focus on multiple objects or parts of interest individually   
 - **Limitation** : the number of parallel transformers limits the number of modeled objects

### 4.1 Distorted MNIST
![image](https://user-images.githubusercontent.com/83633885/136637218-058d1efb-a269-48b7-9b3a-4661988d3388.png)

**[ Train ]**
- Transformation type : Rotation(R), Rotation-Translataion-Scale(RTS), Projective(P), Elastic warping(E)
- Network type : FCN, CNN, ST-FCN, ST-CNN 
- Sampling (bilinear) : affine(Aff), projective(Proj), thin plate spline(TPS)
- Identical condition : same # of params, same base structure, identical optim (backprop, SGD, scheduled lr decay, multinomial CE loss, three weight layers) / CNN includes 2 max-pooling

**[ Result ]**
- Network type : (percent error) **ST-CNN** < ST-FCN < CNN < FCN
  - **ST** < non ST : ST enables network outperform
  - **CNN** < FCN : Max-pooling (more spatial invariance) & Convolutional layer itself (better local structure model)
  - CNN = ST-FCN for RTS : ST is alternative way for spatial invariance   
- Sampling : TPS is the best (elastically deform digits, reduce complexity, not overfit on simple data)
- Transformation of inputs for all ST models : Standard upright posed digit = mean pose found in training data
- https://www.youtube.com/watch?v=Ywv0Xi2-14Y&t=94s

### 4.2 Street View House Numbers (SVHN)
![image](https://user-images.githubusercontent.com/83633885/136637223-8ef5dfcf-b0f1-4fa6-bad6-34dfdc1d0136.png)

**[ Train ]**
- Dataset : 1 and 5 digits house number in real world images (200K) 
- Pre-processing : 64 x 64 crop + additional loosely 128 x 128 crop
- CNN : 11 hidden layers, 5 digit-independent softmax
- ST-CNN : Single (f = 4 layers CNN, following input of baseline CNN) / Multi (f = 2 layer FC, before each first 4 CNN)  -> affine transform, bilinear sampler
- SGD, dropout, randomly initialized weights except for regression layers

**[ Result ]**
- Best Accurcay : **ST-CNN Multi** for 64 x 64 images (3.6% error)
  - crop and rescale by focusing resolution and network capacity only on corresponding parts of digit
- Computation Speed : ST-CNN is only **6% slower** than CNN 
  - ST-CNN requires only a single forward pass 

### 4.3 Fine-Grained Classification
![image](https://user-images.githubusercontent.com/83633885/136637225-6c288361-eeb8-4cb7-8c3b-20647967e7e7.png)

**[ Train ]**
- Dataset : CUB-200-2011 birds dataset (6K train, 5.8K test, 200 species)
- Baseline **CNN : Inception + BN** (pre-trained on ImageNet, fine-tuned on CUB) 
- **ST-CNN** : 2 or 4 parallel spatial transformers
- 1 softmax layer, end-to-end backprop

**[ Result ]**
- Best Accurcay : **4 x ST-CNN** (84.1%) -> outperform baseline CNN (82.3%)
- Pose detection (Attention) : head (red) + central part (green) without any additional supervision
- Same performance even if resolution is downsampled (448px input -> 224px output)

## 5. Conclusion
- can be dropped into a network, perform explicit spatial transformations
- can do end-to-end without any change in loss function
- gain accuracy across multiple task
- regressed transformation parameters are available as output
- Expectation : powerful in recurrent models, object reference frame, 3D transformation

## Code

```python
#### 1) Load Dataset ####

from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim
import torchvision
from torchvision import datasets, transforms
import matplotlib.pyplot as plt
import numpy as np

plt.ion()  
from six.moves import urllib
opener = urllib.request.build_opener()
opener.addheaders = [('User-agent', 'Mozilla/5.0')]
urllib.request.install_opener(opener)

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

# train dataset
train_loader = torch.utils.data.DataLoader(
    datasets.MNIST(root='.', train=True, download=True,
                   transform=transforms.Compose([
                       transforms.ToTensor(),
                       transforms.Normalize((0.1307,), (0.3081,))
                   ])), batch_size=64, shuffle=True, num_workers=4)
# test dataset
test_loader = torch.utils.data.DataLoader(
    datasets.MNIST(root='.', train=False, transform=transforms.Compose([
        transforms.ToTensor(),
        transforms.Normalize((0.1307,), (0.3081,))
    ])), batch_size=64, shuffle=True, num_workers=4)


#### 2) Compose STN ####

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv2d(1, 10, kernel_size=5)
        self.conv2 = nn.Conv2d(10, 20, kernel_size=5)
        self.conv2_drop = nn.Dropout2d()
        self.fc1 = nn.Linear(320, 50)
        self.fc2 = nn.Linear(50, 10)

        # Localization-network
        self.localization = nn.Sequential(
            nn.Conv2d(1, 8, kernel_size=7),
            nn.MaxPool2d(2, stride=2),
            nn.ReLU(True),
            nn.Conv2d(8, 10, kernel_size=5),
            nn.MaxPool2d(2, stride=2),
            nn.ReLU(True)
        )

        # 3 * 2 size Affine matrix에 대해 예측
        self.fc_loc = nn.Sequential(
            nn.Linear(10 * 3 * 3, 32),
            nn.ReLU(True),
            nn.Linear(32, 3 * 2)
        )

        # 항등 변환(identity transformation) -> 가중치/바이어스 초기화
        self.fc_loc[2].weight.data.zero_()
        self.fc_loc[2].bias.data.copy_(torch.tensor([1, 0, 0, 0, 1, 0], dtype=torch.float))

    # STN forward
    def stn(self, x):
        xs = self.localization(x)
        xs = xs.view(-1, 10 * 3 * 3)
        theta = self.fc_loc(xs)
        theta = theta.view(-1, 2, 3)

        grid = F.affine_grid(theta, x.size())
        x = F.grid_sample(x, grid)

        return x

    def forward(self, x):
        x = self.stn(x)

        # general forward pass
        x = F.relu(F.max_pool2d(self.conv1(x), 2))
        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))
        x = x.view(-1, 320)
        x = F.relu(self.fc1(x))
        x = F.dropout(x, training=self.training)
        x = self.fc2(x)
        return F.log_softmax(x, dim=1)

model = Net().to(device)


#### 3) Train and Test ####

 optimizer = optim.SGD(model.parameters(), lr=0.01)

def train(epoch):
    model.train()
    for batch_idx, (data, target) in enumerate(train_loader):
        data, target = data.to(device), target.to(device)

        optimizer.zero_grad()
        output = model(data)
        loss = F.nll_loss(output, target)
        loss.backward()  # End-to-end training
        optimizer.step()
        if batch_idx % 500 == 0:
            print('Train Epoch: {} [{}/{} ({:.0f}%)]\tLoss: {:.6f}'.format(
                epoch, batch_idx * len(data), len(train_loader.dataset),
                100. * batch_idx / len(train_loader), loss.item()))
        
def test():
    with torch.no_grad():
        model.eval()
        test_loss = 0
        correct = 0
        for data, target in test_loader:
            data, target = data.to(device), target.to(device)
            output = model(data)

            test_loss += F.nll_loss(output, target, size_average=False).item()
            pred = output.max(1, keepdim=True)[1]
            correct += pred.eq(target.view_as(pred)).sum().item()

        test_loss /= len(test_loader.dataset)
        print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.0f}%)\n'
              .format(test_loss, correct, len(test_loader.dataset),
                      100. * correct / len(test_loader.dataset)))


#### 4) Visualization ####

def convert_image_np(inp):
    """Convert a Tensor to numpy image."""
    inp = inp.numpy().transpose((1, 2, 0))
    mean = np.array([0.485, 0.456, 0.406])
    std = np.array([0.229, 0.224, 0.225])
    inp = std * inp + mean
    inp = np.clip(inp, 0, 1)
    return inp

def visualize_stn():
    with torch.no_grad():
        data = next(iter(test_loader))[0].to(device)

        input_tensor = data.cpu()
        transformed_input_tensor = model.stn(data).cpu()

        in_grid = convert_image_np(
            torchvision.utils.make_grid(input_tensor))

        out_grid = convert_image_np(
            torchvision.utils.make_grid(transformed_input_tensor))

        f, axarr = plt.subplots(1, 2)
        axarr[0].imshow(in_grid)
        axarr[0].set_title('Dataset Images')

        axarr[1].imshow(out_grid)
        axarr[1].set_title('Transformed Images')

for epoch in range(1, 20 + 1):
    train(epoch)
    test()

visualize_stn()

plt.ioff()
plt.show()

```
![image](https://user-images.githubusercontent.com/83633885/136639310-209001ca-8d05-4070-9835-c03c1b9bd508.png)
